File: backend/src/main/kotlin/com/trackwatch/backend/clients/spotify/SpotifyPlaylistApiClient.kt
package com.trackwatch.backend.clients.spotify

import com.trackwatch.backend.clients.spotify.dto.SpotifyPlaylistTracksResponse
import com.trackwatch.backend.exception.ErrorCode
import com.trackwatch.backend.exception.InternalServerErrorException
import com.trackwatch.backend.model.User
import com.trackwatch.backend.service.MetricService
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class SpotifyPlaylistApiClient(metricService: MetricService) : SpotifyApiClient(metricService) {

    private val log = LoggerFactory.getLogger(this::class.java)

    fun addTracksToPlaylist(user: User, playlistId: String, trackUris: List<String>): Map<*, *> {
        val body = mapOf("uris" to trackUris)

        try {
            sendMetricApiCall("addTracksToPlaylist")

            val response = webClient.post()
                .uri("/playlists/$playlistId/tracks")
                .bodyValue(body)
                .header("Authorization", "Bearer ${user.auth.current.accessToken}")
                .retrieve()
                .bodyToMono(Map::class.java)
                .block() ?: throw InternalServerErrorException(
                ErrorCode.UNHANDLED_EXCEPTION,
                "Failed to add tracks to playlist"
            )

            return response

        } catch (e: Exception) {
            throw InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, e.toString())
        }
    }

    fun getPlaylistTracks(user: User, playlistId: String): List<String> {
        val trackUris = mutableListOf<String>()
        var offset = 0
        val limit = 100

        try {
            do {
                sendMetricApiCall("getPlaylistTracks")

                val response = webClient.get()
                    .uri { uriBuilder ->
                        uriBuilder.path("/playlists/$playlistId/tracks")
                            .queryParam("limit", limit)
                            .queryParam("offset", offset)
                            .build()
                    }
                    .header("Authorization", "Bearer ${user.auth.current.accessToken}")
                    .retrieve()
                    .bodyToMono(SpotifyPlaylistTracksResponse::class.java)
                    .block() ?: throw InternalServerErrorException(
                    ErrorCode.UNHANDLED_EXCEPTION,
                    "Failed to get playlist tracks"
                )

                val uris = mapResponseToTrackUris(response)
                trackUris.addAll(uris)
                offset += limit

            } while (offset < response.total)

            return trackUris

        } catch (e: Exception) {
            throw InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, e.toString())
        }
    }

    fun filterSavedTracks(user: User, uris: Set<String>): Set<String> {
        val ids = uris.map { it.split(":").last() }
        val filteredUris = mutableSetOf<String>()
        val limit = 50

        try {
            ids.chunked(limit).forEach { chunk ->
                val idsParam = chunk.joinToString(",")

                sendMetricApiCall("filterSavedTracks")
                val response = webClient.get()
                    .uri { uriBuilder ->
                        uriBuilder.path("/me/tracks/contains")
                            .queryParam("ids", idsParam)
                            .build()
                    }
                    .header("Authorization", "Bearer ${user.auth.current.accessToken}")
                    .retrieve()
                    .bodyToMono(List::class.java)
                    .block() ?: throw InternalServerErrorException(
                    ErrorCode.UNHANDLED_EXCEPTION,
                    "Failed to check saved tracks"
                )

                val savedStatuses = response as List<Boolean>

                chunk.forEachIndexed { index, id ->
                    if (!savedStatuses[index]) {
                        filteredUris.add("spotify:track:$id")
                    }
                }
            }

        } catch (e: Exception) {
            throw InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, e.toString())
        }

        return filteredUris
    }

    fun createPlaylist(user: User, name: String, description: String, isPublic: Boolean): String {
        val body = mapOf(
            "name" to name,
            "description" to description,
            "public" to isPublic
        )

        try {
            sendMetricApiCall("createPlaylist")

            val response = webClient.post()
                .uri("/users/${user.id}/playlists")
                .bodyValue(body)
                .header("Authorization", "Bearer ${user.auth.current.accessToken}")
                .retrieve()
                .bodyToMono(Map::class.java)
                .block() ?: throw InternalServerErrorException(
                ErrorCode.UNHANDLED_EXCEPTION,
                "Failed to create playlist"
            )

            return response["id"] as String

        } catch (e: Exception) {
            throw InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, e.toString())
        }
    }

    fun checkPlaylistExistsWithRetries(user: User, maxRetries: Int = 5, delayMs: Long = 500): Boolean {
        repeat(maxRetries) {
            if (checkPlaylistExists(user)) return true
            Thread.sleep(delayMs)
        }

        return false
    }

    fun checkPlaylistExists(user: User): Boolean {
        val userPlaylistId = user.playlistId

        var offset = 0
        val limit = 50

        val allPlaylists = mutableListOf<Map<String, Any>>()

        try {
            do {
                sendMetricApiCall("checkPlaylistExists")
                val response = webClient.get()
                    .uri { uriBuilder ->
                        uriBuilder.path("/me/playlists")
                            .queryParam("limit", limit)
                            .queryParam("offset", offset)
                            .build()
                    }
                    .header("Authorization", "Bearer ${user.auth.current.accessToken}")
                    .retrieve()
                    .bodyToMono(Map::class.java)
                    .block() ?: throw InternalServerErrorException(
                    ErrorCode.UNHANDLED_EXCEPTION,
                    "Failed to get user playlists"
                )

                val items = response["items"] as List<*>
                collectPlaylistItems(items, allPlaylists)

                val playlistIds = items.map { (it as Map<*, *>)["id"] as String }
                if (playlistIds.contains(userPlaylistId)) return true

                offset += limit
                val total = (response["total"] as Int?) ?: 0
            } while (offset < total)

            logAvailablePlaylists(userPlaylistId, allPlaylists)
            return false
        } catch (e: Exception) {
            throw InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, e.toString())
        }
    }

    fun updatePlaylistCover(user: User, playlistId: String, imageBase64: String) {
        try {
            sendMetricApiCall("updatePlaylistCover")

            webClient.put()
                .uri("/playlists/$playlistId/images")
                .header("Authorization", "Bearer ${user.auth.current.accessToken}")
                .header("Content-Type", "image/jpeg")
                .bodyValue(imageBase64)
                .retrieve()
                .bodyToMono(Void::class.java)
                .block()

        } catch (e: Exception) {
            throw InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, e.toString())
        }
    }

    private fun collectPlaylistItems(items: List<*>, allPlaylists: MutableList<Map<String, Any>>) {
        items.forEach {
            val playlist = it as Map<*, *>
            allPlaylists.add(mapOf(
                "id" to (playlist["id"] as String),
                "name" to (playlist["name"] as String)
            ))
        }
    }

    private fun mapResponseToTrackUris(response: SpotifyPlaylistTracksResponse): List<String> {
        return response.items.mapNotNull { it.track?.uri }
    }

    private fun logAvailablePlaylists(searchedPlaylistId: String, playlists: List<Map<String, Any>>) {
        log.info("Playlist with ID: $searchedPlaylistId not found. Available playlists:")
        playlists.forEach { playlist ->
            log.info("ID: ${playlist["id"]}, Name: ${playlist["name"]}")
        }
    }

}

File: backend_py/trackwatch/app/clients/spotify/spotify_playlist_api_client.py
from .spotify_api_client import spotify_api_request
from app.exceptions import InternalServerErrorException, ErrorCode
import time

def add_tracks_to_playlist(user, playlist_id, track_uris):
  body = {"uris": track_uris}
  try:
    response = spotify_api_request(
      method="POST",
      endpoint=f"/playlists/{playlist_id}/tracks",
      token=user.current_access_token,
      json_data=body
    )
    return response
  except Exception as e:
    raise InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, str(e))

def get_playlist_tracks(user, playlist_id):
  track_uris = []
  offset = 0
  limit = 100
  try:
    while True:
      response = spotify_api_request(
        method="GET",
        endpoint=f"/playlists/{playlist_id}/tracks",
        token=user.current_access_token,
        params={"limit": limit, "offset": offset}
      )
      uris = map_response_to_track_uris(response)
      track_uris.extend(uris)
      offset += limit
      total = response.get("total", 0)
      if offset >= total:
        break
    return track_uris
  except Exception as e:
    raise InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, str(e))

def filter_saved_tracks(user, uris):
  ids = [uri.split(":")[-1] for uri in uris]
  filtered_uris = []
  limit = 50

  try:
    for i in range(0, len(ids), limit):
      chunk = ids[i:i+limit]
      ids_param = ",".join(chunk)

      response = spotify_api_request(
        method="GET",
        endpoint="/me/tracks/contains",
        token=user.current_access_token,
        params={"ids": ids_param}
      )
      saved_statuses = response

      for idx, id_ in enumerate(chunk):
        if not saved_statuses[idx]:
          filtered_uris.append(f"spotify:track:{id_}")

    return filtered_uris
  except Exception as e:
    raise InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, str(e))

def create_playlist(user, name, description, is_public):
  body = {
    "name": name,
    "description": description,
    "public": is_public
  }
  try:
    response = spotify_api_request(
      method="POST",
      endpoint=f"/users/{user.id}/playlists",
      token=user.current_access_token,
      json_data=body
    )
    return response["id"]
  except Exception as e:
    raise InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, str(e))

def check_playlist_exists_with_retries(user, max_retries=5, delay_s=0.5):
  for _ in range(max_retries):
    if check_playlist_exists(user):
      return True
    time.sleep(delay_s)
  return False

def check_playlist_exists(user):
  user_playlist_id = user.playlist_id
  offset = 0
  limit = 50
  all_playlists = []
  try:
    while True:
      response = spotify_api_request(
        method="GET",
        endpoint="/me/playlists",
        token=user.current_access_token,
        params={"limit": limit, "offset": offset}
      )
      items = response.get("items", [])
      collect_playlist_items(items, all_playlists)
      playlist_ids = [pl["id"] for pl in items if isinstance(pl, dict)]
      if user_playlist_id in playlist_ids:
        return True
      offset += limit
      total = response.get("total", 0)
      if offset >= total:
        break
    return False
  except Exception as e:
    raise InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, str(e))

def update_playlist_cover(user, playlist_id, image_base64):
  try:
    spotify_api_request(
      method="PUT",
      endpoint=f"/playlists/{playlist_id}/images",
      token=user.current_access_token,
      data=image_base64,
      headers={"Content-Type": "image/jpeg"}
    )
  except Exception as e:
    print(f"Error updating playlist cover: {e}")
    # raise InternalServerErrorException(ErrorCode.UNHANDLED_EXCEPTION, str(e))

def collect_playlist_items(items, all_playlists):
  for playlist in items:
    if isinstance(playlist, dict):
      all_playlists.append({
        "id": playlist.get("id"),
        "name": playlist.get("name")
      })

def map_response_to_track_uris(response):
  items = response.get("items", [])
  uris = []
  for item in items:
    track = item.get("track") if isinstance(item, dict) else None
    if track and track.get("uri"):
      uris.append(track["uri"])
  return uris
